### ISSUE TITLE
TEE security consideration

### ISSUE THREAD

**Issue by gselander** [lake-wg/edhoc#99]
Section 8.6 in -05 states:

  "Typically an 
    adversary with physical access to a device can be assumed to gain 
    access to all information outside of the TEE, but none of the 
    information inside the TEE."

Stefan Hristozov commented:

    "A TEE does not provide protection against physical adversary, since the same CPU with the same physical memory executes both the operations in the TEE and in the non-TEE domains. It provides protection only against remote adversary exploiting, e.g., a buffer overflow for conducting code injection or return oriented attacks."

Following the description in [1] I replaced the text with:

"The use of a TEE enforces that code within that environment cannot be tampered with, and that any data used by such code cannot be read or tampered with by code outside that environment."

https://github.com/lake-wg/edhoc/commit/665d60c

Any objections to this?

[1] https://tools.ietf.org/html/draft-ietf-teep-architecture-14

**Comment by emanjon** [comment]
Good to talk about attacks from code outside of the TEE, but not good to remove everything about physical attacks. The original text was written only thinking about physical attacks using the TEE definition from [1]. That definition in [1]states: "The TEE resists against all software attacks as well as the physical attacks performedon the main memory of the system." so I don't think "A TEE does not provide protection against physical adversary" is completely correct. I would prefer the definition of TEE in [1] but I guess I have to accept the that current commercialy avilable TEEs does not enforce does not enforce memory encryption and also only provides a single zone inside the TEE.

I think the suggested new text is ok, but I think it needs to be augmented with a sentence about memory encryption as well as attacks from non-EDHOC code inside the TEE.

**Comment by emanjon** [comment]
I added some additional text on additional attacks and how to protect. I think the new is very good. The TEE definition taken from TEEP describe a quite basic TEE which is good.

If you agree with the updated text we can close this issue.

**Comment by emanjon** [comment]
@StefanHri are you ok with the suggested update?

https://github.com/lake-wg/edhoc/commit/57ac5116f91b81a1e9ada0d157d896eca814d331

**Comment by StefanHri** [comment]
Hi,
I don't understand what you want to say with the sentence "The TEE can also be used to protect the EDHOC and application protocol (e.g. OSCORE) implementation using some form of "secure boot", memory protection etc." Secure boot and TEE are different things. I think both make sense but they are not necessarily related to each other. 

I also think the sentence "To provide better protection against physical attacks, memory encryption needs to be enforced."
is incorrect since physical attacks are a very broad topic and memory encryption certainly does not protect against some common form of physical attacks such as (deferential) power analysis and fault injections.

**Comment by emanjon** [comment]
The "secure boot" text was old, I don't know where it came from. I wrote a new sentence about secure boot and REE.

Regarding physical attacks. bus encryption seems to be recommended to protect against some physical attacks. I agree that this does not protect against all physical attacks.

https://www.usenix.org/system/files/raid2019-zhao.pdf
https://people.dsv.su.se/~icss-jpc/exjobb/interesting/2004_vlsid_invited.pdf

I changed the text to say "some forms of physical attacks" and rewrote the text a bit.

"If supported by the device, it is RECOMMENDED that at least the long-term private keys is stored in a Trusted Execution Environment (TEE) and that sensitive operations using these keys are performed inside the TEE. To achieve even higher security it is RECOMMENDED that additional operations such as ephemeral key generation, all computations of shared secrets, and storage of the PRK keys can be done inside the TEE. The use of a TEE enforces that code within that environment cannot be tampered with, and that any data used by such code cannot be read or tampered with by code outside that environment. Note that non-EDHOC code inside the TEE might still be able to read EDHOC data and tamper with EDHOC code, to protect against such attacks EDHOC needs to be in its own zone. **To provide better protection against some forms of physical attacks, sensitive EDHOC data should be stored inside the SoC or encrypted and integrity protected when send on a data bus (e.g. between the CPU and RAM or Flash). Secure boot can be used to increase the security of code and data in the Rich Execution Environment (REE) by validating the REE image.**"

**Comment by StefanHri** [comment]
There are many different attacks and different countermeasures which depend a lot on the hardware platform. To give an in-depth analysis of different attacks and countermeasures is certainly outside the scope of the standard. I will suggest keeping it short and reference relevant papers where more concrete analysis is done. Moreover, this is a general problem that every security protocol has.

**Comment by emanjon** [comment]
I agree, lets not talk about memory, data buses, and secure boot at all. I removed all the old and new text regarding these things. I don't know where they came from originally. After the last commit the text equal to the -06 version but with the sentence on secure boot and memory protection deleted.
